src/components/LoginForm.js

```diff
@@ -2,80 +2,150 @@ import React from 'react';
+                  </div>
+                </div>
+                <div className="form-group">
+                  <label htmlFor="spPassword" className="col-xs-12 col-sm-4 control-label">Password</label>
+                  <div className="col-xs-12 col-sm-4">
+                    <input type="password" className="form-control" id="spPassword" name="password" placeholder="Password" />
+                  </div>
+                </div>
+                <div className="form-group">
+                  <div className="col-sm-offset-4 col-sm-4">
+                    <p className="alert alert-danger" spIf="form.error"><span spBind="form.errorMessage" /></p>
+                    <button type="submit" className="btn btn-primary">Login</button>
+                    <Link to="/forgot" className="pull-right">Forgot Password</Link>
+                  </div>
+                </div>
+              </div>
+            </div>
+          </div>
+        </div>
+      </LoginForm>
+    );
+  }
+}
+
 @ReactMixin.decorate(History)
 export default class LoginForm extends React.Component {
   state = {
-    username: '',
-    password: '',
-    isProcessing: false,
-    errorMessage: null
+    fields: {
+      username: '',
+      password: '',
+    },
+    errorMessage: null,
+    isFormProcessing: false
   };
 
   onFormSubmit(e) {
     e.preventDefault();
 
-    var self = this;
+    var next = (err, data) => {
+      if (err) {
+        return this.setState({
+          isFormProcessing: false,
+          errorMessage: err.message
+        });
+      }
 
-    var redirectTo = this.props.redirectTo || '/';
+      // If the user didn't specify any data,
+      // then simply default to what we have in state.
+      data = data || this.state.fields;
 
-    self.setState({ isProcessing: true });
+      UserActions.login({
+        username: data.username,
+        password: data.password
+      }, (err, result) => {
+        this.setState({ isFormProcessing: false });
 
-    UserActions.login({
-      username: this.state.username,
-      password: this.state.password
-    }, function (err, result) {
-      self.setState({ isProcessing: false });
+        if (err) {
+          this.setState({ errorMessage: err.message });
+        } else {
+          this.history.pushState(null, this.props.redirectTo || '/');
+        }
+      });
+    };
 
-      if (err) {
-        self.setState({ errorMessage: err.message });
-      } else {
-        self.history.pushState(null, redirectTo);
-      }
+    this.setState({
+      isFormProcessing: true
     });
+
+    if (this.props.onSubmit) {
+      e.data = this.state.fields;
+      this.props.onSubmit(e, next);
+    } else {
+      next(null, this.state.fields);
+    }
   }
 
-  onUsernameChanged(e) {
-    this.setState({ username: e.target.value });
+  _mapFormFieldHandler(element, tryMapField) {
+    if (['input', 'textarea'].indexOf(element.type) > -1) {
```

Comment:
I don't agree with this since right now the code is isolated in a block. Also, you've inverted the statements which in my opinion makes it harder to read. But what I'm going to change is `['input', 'textarea'].indexOf(element.type) > -1` into `element.type === 'input' || element.type === 'textarea'`, since that actually makes it easier to read.